src/game/AI/ScriptDevAI/scripts/northrend/naxxramas/boss_sapphiron.cpp

```diff
@@ -16,339 +16,398 @@
+    SPELL_SUMMON_BLIZZARD       = 28561,
+
+    // Air phase spells
+    SPELL_DRAGON_HOVER          = 18430,
+    SPELL_ICEBOLT_INIT          = 28526,            // Triggers spell 28522 (Icebolt)
+    SPELL_ICEBOLT               = 28522,
+    SPELL_ICEBOLT_IMMUNITY      = 31800,
+    SPELL_ICEBLOCK_SUMMON       = 28535,
+    SPELL_FROST_BREATH_DUMMY    = 30101,
+    SPELL_FROST_BREATH          = 28524,            // Triggers spells 29318 (Frost Breath) and 30132 (Despawn Ice Block)
+    SPELL_DESPAWN_ICEBLOCK_GO   = 28523,
+    SPELL_SUMMON_WING_BUFFET    = 29329,
+    SPELL_DESPAWN_WING_BUFFET   = 29330,            // Triggers spell 29336 (Despawn Buffet)
+    //SPELL_WING_BUFFET           = 29328,
+    SPELL_DESPAWN_BUFFET_EFFECT = 29336,
     SPELL_CHILL                 = 28547,
     SPELL_CHILL_H               = 55699,
-    SPELL_SUMMON_BLIZZARD       = 28560,
-    SPELL_BESERK                = 26662,
     SPELL_ACHIEVEMENT_CHECK     = 60539,                    // unused
 
-    NPC_YOU_KNOW_WHO            = 16474,
+    NPC_BLIZZARD                = 16474,
 };
 
 static const float aLiftOffPosition[3] = {3522.386f, -5236.784f, 137.709f};
 
-enum Phases
+enum SapphironPhases
 {
     PHASE_GROUND        = 1,
     PHASE_LIFT_OFF      = 2,
-    PHASE_AIR_BOLTS     = 3,
-    PHASE_AIR_BREATH    = 4,
-    PHASE_LANDING       = 5,
 };
 
-struct boss_sapphironAI : public ScriptedAI
+enum SapphironActions
+{
+    SAPPHIRON_CLEAVE,
+    SAPPHIRON_TAIL_SWEEP,
+    SAPPHIRON_ICEBOLT,
+    SAPPHIRON_FROST_BREATH,
+    SAPPHIRON_LIFE_DRAIN,
+    SAPPHIRON_BLIZZARD,
+    SAPPHIRON_AIR_PHASE,
+    SAPPHIRON_LANDING_PHASE,
+    SAPPHIRON_GROUND_PHASE,
+    SAPPHIRON_BERSERK,
+    SAPPHIRON_ACTION_MAX,
+};
+
+std::unordered_set<ObjectGuid> localIceBlocks;
+
+static const uint32 groundPhaseActions[] = {SAPPHIRON_CLEAVE, SAPPHIRON_TAIL_SWEEP, SAPPHIRON_BLIZZARD, SAPPHIRON_LIFE_DRAIN};
+
+struct boss_sapphironAI : public CombatAI
 {
-    boss_sapphironAI(Creature* pCreature) : ScriptedAI(pCreature)
+    boss_sapphironAI(Creature* creature) : CombatAI(creature, SAPPHIRON_ACTION_MAX), m_instance(static_cast<ScriptedInstance*>(creature->GetInstanceData()))
     {
-        m_pInstance = (instance_naxxramas*)pCreature->GetInstanceData();
-        m_bIsRegularMode = pCreature->GetMap()->IsRegularDifficulty();
-        Reset();
+        AddCombatAction(SAPPHIRON_CLEAVE, 5u * IN_MILLISECONDS);
+        AddCombatAction(SAPPHIRON_TAIL_SWEEP, 12u * IN_MILLISECONDS);
+        AddCombatAction(SAPPHIRON_LIFE_DRAIN, 11u * IN_MILLISECONDS);
+        AddCombatAction(SAPPHIRON_BLIZZARD, 15u * IN_MILLISECONDS);
+        AddCombatAction(SAPPHIRON_FROST_BREATH, true);
+        AddCombatAction(SAPPHIRON_ICEBOLT, true);
+        AddCombatAction(SAPPHIRON_BERSERK, 15u * MINUTE * IN_MILLISECONDS);
+        AddCustomAction(SAPPHIRON_AIR_PHASE, true, [&]() { HandleAirPhase(); });
+        AddCustomAction(SAPPHIRON_LANDING_PHASE, true, [&]() { HandleLandingPhase(); });
```

Comment:
Make sure you realize what making the transition a custom action means. It means that it can go off during a spellcast, which is generally wrong. NPCs generally finish what they were doing instead of interrupting to do a phase transition.

Reply:
In this case it should be OK, because Sapphiron isn't casting anything, only waiting for his previous cast to impact. What he does while in the air is tightly defined. No random casts or anything happen while he's airborne

Reply:
Well that just seems incorrect from me reading the code.  void HandleAirPhase() does a InterruptNonMeleeSpells
which is exactly the general problem. That code should not be there and it shouldnt go off if there were a cast happening. I get that ground phase sapphiron does not have spells with cast time, but this is a general rule in any script. "Finish your current action, and then proceed" Which is generically guaranteed by it being a combat action.

Reply:
I can rewrite them into CombatActions easy enough. Fair enough, I didn't consider the general case, because I know Sapphiron doesn't cast anything with cast-time